modules/perception/obstacle/camera/filter/object_camera_extended_kalman_filter.cc

```diff
@@ -0,0 +1,198 @@
+/******************************************************************************
+ * Copyright 2018 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/perception/obstacle/camera/filter/object_camera_extended_kalman_filter.h"
+
+#include <cmath>
+#include <string>
+#include <vector>
+
+namespace apollo {
+namespace perception {
+
+using common::math::ExtendedKalmanFilter;
+using ObjectFilter = ObjectCameraExtendedKalmanFilter::ObjectFilter;
+
+void ObjectCameraExtendedKalmanFilter::GetState(const int track_id,
+    const std::shared_ptr<VisualObject>& obj_ptr) {
+
+  const auto& filter = tracked_filters_[track_id].object_config_filter_;
+  Eigen::Vector4f mean = filter.GetStateMean();
+  Eigen::Matrix4f covariance = filter.GetStateCovariance();
+
+  obj_ptr->center.x() = mean[0];
+  obj_ptr->velocity.x() = mean[3] * std::cos(mean[2]);
+
+  obj_ptr->center.y() = mean[1];
+  obj_ptr->velocity.y() = mean[3] * std::sin(mean[2]);
+
+  obj_ptr->center.z() = 0.0f;
+  obj_ptr->velocity.z() = 0.0f;
+
+  obj_ptr->state_uncertainty.block(0, 0, 2, 2) << covariance(0, 0),
+      covariance(0, 1), covariance(1, 0), covariance(1, 1);
+  // TODO(zhangyajia): set the covariance for other variables
+
+  obj_ptr->theta = mean[2];
+  obj_ptr->direction = Eigen::Vector3f(std::cos(obj_ptr->theta), 0.0f,
+      -std::sin(obj_ptr->theta));
+}
+
+bool ObjectCameraExtendedKalmanFilter::Filter(const double timestamp,
+    std::vector<std::shared_ptr<VisualObject> >* objects) {
+  if (!objects) {
+    return false;
+  }
+
+  // update lost_frame_count
+  for (auto& p : tracked_filters_) {
+    p.second.lost_frame_cnt_ += 1;
+  }
+
+  for (auto& obj_ptr : *objects) {
+    int track_id = obj_ptr->track_id;
+
+    if (track_id != -1 && tracked_filters_.count(track_id) > 0) {
+      Predict(track_id, timestamp);
+      Update(track_id, obj_ptr);
+      GetState(track_id, obj_ptr);
+    } else {
+      tracked_filters_[track_id] = CreateObjectFilter(track_id, timestamp,
+          obj_ptr);
+    }
+  }
+
+  Destroy();
+  return true;
+}
+
+void ObjectCameraExtendedKalmanFilter::Predict(const int track_id,
+    const double timestamp) {
+  double time_diff = timestamp - tracked_filters_[track_id].last_timestamp_;
+
+  // update transition model
```

Comment:
@YajiaZhang You probably want to make a note here that this camera EKF is mainly for vehicle kinematics?

Reply:
@YajiaZhang Also make a note saying x[0], x[1] is obstacle position in FLU, x[2] is theta heading angle, x[3] is velocity?

Reply:
Not necessary. Although we assume the obstacles are vehicles, this works for non vehicles as well.